src/expr/src/scalar/func.rs

```diff
@@ -1750,6 +1755,49 @@ fn date_part_interval_inner(
     }
 }
 
+fn date_part_time<'a, T>(a: Datum<'a>, time: T) -> Result<Datum<'a>, EvalError>
+where
+    T: TimeLike
+{
+    let units = a.unwrap_str();
+    match units.parse() {
+        Ok(units) => date_part_time_inner(units, time),
+        Err(_) => Err(EvalError::UnknownUnits(units.to_owned())),
+    }
+}
+
+fn date_part_time_inner<'a, T>(units: DateTimeUnits, time: T) -> Result<Datum<'a>, EvalError>
+where
+    T: TimeLike
+{
+    match units {
+        DateTimeUnits::Hour => Ok(time.extract_hour().into()),
+        DateTimeUnits::Minute => Ok(time.extract_minute().into()),
+        DateTimeUnits::Second => Ok(time.extract_second().into()),
+        DateTimeUnits::Milliseconds => Ok(time.extract_millisecond().into()),
+        DateTimeUnits::Microseconds => Ok(time.extract_microsecond().into()),
+        DateTimeUnits::Epoch
+        | DateTimeUnits::Year
+        | DateTimeUnits::Week
+        | DateTimeUnits::Day
+        | DateTimeUnits::Millennium
+        | DateTimeUnits::Century
+        | DateTimeUnits::Decade
+        | DateTimeUnits::Quarter
+        | DateTimeUnits::Month
+        | DateTimeUnits::DayOfWeek
+        | DateTimeUnits::DayOfYear
+        | DateTimeUnits::IsoDayOfWeek
+        | DateTimeUnits::IsoDayOfYear => Err(EvalError::UnknownUnits(format!("{}", units))),
```

Comment:
I think including the type for which these are unrecognized units (i.e. matching PG) would be nice here because this represents the first type with units disjoint from the others. As Nikhil points out elsewhere, not a big deal to match things _so_ specifically, but this would be a nice piece of UX polish no one is likely to do at a later point in time.

Reply:
I found yet another inconsistency in Postgres error messages, sometimes they say "not supported" instead of "not recognized". Ex:
```
postgres=# SELECT EXTRACT(seccoond FROM DATE '2006-01-01');
ERROR:  date units "seccoond" not recognized
postgres=# SELECT EXTRACT(second FROM DATE '2006-01-01');
ERROR:  date units "second" not supported
```
I decided to add this as a separate error message because I thought that it conveyed more clearly that the unit type is known, just not supported for the given type.

Reply:
Yeah turns out it's an intentional deviation, per Tom Lane: https://www.postgresql.org/message-id/flat/CAPWqQZRa58atSHDgbDraAAo2371zBdmcAej7e-%3D8g%2BYxJa35uA%40mail.gmail.com#cd1c7e5dc7c7ee8364920af6cc1801ab. Since they use different error codes using distinct error variants is exactly right.

If that patch goes in then the error messages will change to:

    unit "second" not supported for type date

and:

    unit "second" not recognized for type date

But feel free to dump any of that work into #9853 if you want to get on with your life. :)

Reply:
I'll include it in this PR, since it's so small. Though for the not recognized error we don't have a good way of deriving the type in the general case so I'll leave it out for now. 

As a follow up I noticed some other inconsistencies in our date-time functions from Postgres. I'm planning on detailing them in an issue after this.

Reply:
Also if your interested in some more postgres inconsistencies, here's an interesting one I noticed related to this PR. The date-time docs can be found here: https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT


As per the `EXTRACT` docs, `EXTRACT` is implemented for `time`, so you cannot try and extract a date related field from `time`.
> source must be a value expression of type timestamp, time, or interval. (Expressions of type date are cast to timestamp and can therefore be used as well.)
```
postgres=# SELECT EXTRACT(month FROM time '20:38:40');
ERROR:  "time" units "month" not recognized
```


As per the `DATE_TRUNC` docs, `DATE_TRUNC` is not implemented for `time`. Instead `time` gets cast to `interval`. Since date related fields are valid for `interval`, you can trunc a `time` using date related fields.
> source is a value expression of type timestamp, timestamp with time zone, or interval. (Values of type date and time are cast automatically to timestamp or interval, respectively.) 
```
postgres=# SELECT DATE_TRUNC('month', time '20:38:40');
 date_trunc 
------------
 00:00:00
(1 row)
```


As per the `EXTRACT` docs, `EXTRACT` is not implemented for `date`. Instead `date` gets cast to `timestamps`. Using the logic before, since time related fields are valid for `timestamp`, then you should be able to extract time related fields from `date`.
> source must be a value expression of type timestamp, time, or interval. (Expressions of type date are cast to timestamp and can therefore be used as well.)
```
postgres=# SELECT EXTRACT(second FROM date '2001-02-16');
ERROR:  date units "second" not supported
```
However as you can see you can't.


As per the `DATE_TRUNC` docs, `DATE_TRUNC` is not implemented for `date`. Instead `date` gets cast to `timestamp`. Now using the logic before it's unclear whether or not you can trunc a `date` using time related fields.
>  source is a value expression of type timestamp, timestamp with time zone, or interval. (Values of type date and time are cast automatically to timestamp or interval, respectively.) 
```
postgres=# SELECT DATE_TRUNC('second', date '2001-02-16');
       date_trunc       
------------------------
 2001-02-16 00:00:00-05
(1 row)
```
However it seems like you can.


Not really sure how to reconcile these inconsistencies, maybe it's just a bug.

Reply:
Mystery solved. The documentation seems to be incorrect. `EXTRACT` actually is implemented for the `date` type (https://github.com/postgres/postgres/blob/dfe67c0e85a5a613f802f78641ccc1c48845076e/src/backend/utils/adt/date.c#L1068-L1242). So it validates the units.

`DATE_PART` however is not implemented for the `date` type. so it doesn't validate the units.
```
postgres=#  SELECT EXTRACT(second FROM date '2000-12-16 12:21:13');
ERROR:  date units "second" not supported

postgres=#  SELECT DATE_PART('second', date '2000-12-16 12:21:13');
 date_part 
-----------
         0
(1 row)

```